copt/randomized.py

```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+          the optimizer exited successfully and ``message`` which describes
+          the cause of the termination. See `scipy.optimize.OptimizeResult`
+          for a description of other attributes.
+
+     References:
+
+    .. [NDTELP2020] Negiar, Geoffrey, Dresdner, Gideon, Tsai Alicia, El Ghaoui, Laurent, Locatello, Francesco, and Pedregosa, Fabian.
+    `"Stochastic Frank-Wolfe for Constrained Finite-Sum Minimization" <https://arxiv.org/abs/2002.11860v2>` arxiv:2002.11860v2 (2020).
+
+    .. [MHK2018] Mokhtari, Aryan, Hassani, Hamed, and Karbassi, Amin `"Stochastic Conditional Gradient Methods:
+From Convex Minimization to Submodular Maximization" <https://arxiv.org/abs/1804.09554>`_, arxiv:1804.09554 (2018)
+
+    .. [LF2020] Lu, Haihao, and Freund, Robert `"Generalized Stochastic Frank-Wolfe Algorithm with Stochastic 'Substitute' Gradient for Structured Convex Optimization"
+    <https://arxiv.org/pdf/1806.05123.pdf>`_, Mathematical Programming (2020).
+
+
+    """
+    n_samples, n_features = A.shape
+    x = np.reshape(x0, n_features).astype(float)
+    assert x.shape == (n_features,)
+    A = sparse.csr_matrix(A).copy()
+
+    dual_var = np.zeros(n_samples)
+    grad_agg = np.zeros(n_features)
+
+    if variant == 'LF':
+        agg = utils.safe_sparse_dot(A, x)
+
+    success = False
+
+    if callback is not None:
+        callback(locals())
+
+    if step_size is None:
+        # then default according to variant
+        step_size = step_size_sfw(variant)
+
+    for it in range(max_iter):
+        x_snapshot = x.copy()
+        # Batch size 1
+        idx = randint(0, n_samples - 1)
+
+        step_size_x, step_size_agg = step_size(it, n_samples)
+        dual_var_prev = dual_var[idx]
+
+        if variant in {'SAG', 'SAGA'}:
+            p = A[idx].dot(x)
+            dual_var[idx] = (1 / n_samples) * f_deriv(p, b[idx])
+
+        elif variant == 'MK':
+            p = A[idx].dot(x)
+            dual_var[idx] += step_size_agg * (f_deriv(p, b[idx]) - dual_var[idx])
+
+        elif variant == 'LF':
+            update_direction, _ = lmo(-grad_agg, x)
+            agg[idx] += step_size_agg * (utils.safe_sparse_dot(A[idx], update_direction + x) - agg[idx])
+            dual_var[idx] = (1 / n_samples) * f_deriv(agg[idx], b[idx])
+
+        # For all variants, update the aggregate gradient
+        grad_agg = utils.safe_sparse_add(grad_agg, (dual_var[idx] - dual_var_prev) * A[idx])
+
+        if variant in {'SAG', 'MK'}:
+            update_direction, _ = lmo(-grad_agg, x)
+
+        elif variant == 'SAGA':
+            grad_est = utils.safe_sparse_add(grad_agg, (n_samples - 1) * (dual_var[idx] - dual_var_prev) * A[idx])
+            update_direction, _ = lmo(-grad_est, x)
+
+        x += step_size_x * update_direction
+
+        if callback is not None:
+            callback(locals())
+
+        if np.abs(x - x_snapshot).sum() < tol:
+            success = True
+            break
+    message = ""
```

Comment:
Not familiar with scipy optimize, but are these messages often used? And for what? In this file it seems that it is always empty string, so I guess it's fine.

Reply:
Not sure, this was in other stochastic methods from the file that I was using as template. @fabianp ?